src/dmd/semantic3.d

```diff
@@ -1295,6 +1295,75 @@ private extern(C++) final class Semantic3Visitor : Visitor
             sc = sc.pop();
         }
 
+        // Check `extern(C++)` functions for invalid the return/parameter types
+        if (funcdecl._linkage == LINK.cpp)
+        {
+            static bool checkCppNonMappableType(Type type, Parameter param = null, Type origType = null)
+            {
+                if (origType is null)
+                {
+                    origType = type;
+                    type = type.toBasetype();
+                }
+
+                // Don't allow D `immutable` and `shared` types to be interfaced with C++
+                if (type.isImmutable() || type.isShared())
+                    return true;
+                else if (Type cpptype = target.cpp.parameterType(type))
+                    type = cpptype;
+
+                // Permit types that are handled by toCppMangle.
+                // This list should be kept in sync with dmd.cppmangle and dmd.cppmanglewin.
```

Comment:
Maybe it would be better to put this list in a public function that is used by cppmangle, cppmanglewin and this function? That way we don't need to modify the list in multiple places?

Reply:
It's visitor functions in cppmangle that this should be in sync with.  Nothing is actually duplicated.